src/types/cron.types.ts

```diff
@@ -27,21 +33,35 @@ export type CronJobParams =
 				  }
 			);
 
-export type CronCommand =
-	/**
-	 * TODO: find out how to type the context correctly, based on
-	 * if the "context" was provided to the CronJob constructor
-	 * leaving "any" for now...
-	 */
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	| ((this: CronJob | any) => void)
+export type CronContext<C> = C extends null ? CronJob<null> : NonNullable<C>;
+
+export type CronCallback<C, WithOnCompleteBool extends boolean = false> = (
+	this: CronContext<C>,
+	onComplete: WithOnCompleteBool extends true
+		? OmitThisParameter<CronOnCompleteCallback<C>>
+		: never
+) => void;
+
+export type CronOnCompleteCallback<C> = (this: CronContext<C>) => void;
+
+export type CronSystemCommand =
 	| string
 	| {
 			command: string;
 			args?: readonly string[] | null;
 			options?: SpawnOptions | null;
 	  };
 
+export type CronCommand<C, WithOnCompleteBool extends boolean = false> =
+	| CronCallback<C, WithOnCompleteBool>
+	| CronSystemCommand;
+
+export type CronOnCompleteCommand<C> =
+	| OmitThisParameter<CronOnCompleteCallback<C>>
+	| CronSystemCommand;
+
+export type WithOnComplete<OC> = OC extends null ? false : true;
```

Comment:
just double checking, this means if the `onComplete` callback is provided, the type `WithOnComplete` must be `true` and it's `false` otherwise? seems more like program logic rather than type logic

Reply:
yes that's right, this is used to determine other types around the codebase (to avoid duplicating `OC extends null ? false : true` everywhere)